modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1325 @@
+    const connectedClient = connect("google.com", "http", IPFamily.IPv4, timeout);
+
+  :arg host: host to connect to or resolve if not in standard ip notation
+  :type host: `string`
+  :arg service: service to connect to on resolved `host`
+  :type service: `string`
+  :arg family: type of socket family to connect to
+  :type family: :type:`IPFamily`
+  :arg timeout: time to wait for each possible connection.
+  :type timeval: :type:`~Sys.timeval`
+  :return: connected socket.
+  :rtype: `tcpConn`
+  :throws SystemError: Upon failure to resolve address or connect
+                        to any of the resolved address in given `timeout`.
+*/
+proc connect(in host: string, in service: string, family: IPFamily = IPFamily.IPUnspec, in timeout = new timeval(-1,0)): tcpConn throws {
+  var result:sys_addrinfo_ptr_t;
+  var hints = new sys_addrinfo_t();
+  hints.ai_family = family:c_int;
+  hints.ai_socktype = SOCK_STREAM;
+  var err = getaddrinfo(host.c_str(), service.c_str(), hints, result);
+  if err != 0 {
+    throw new Error("Can't resolve address");
+  }
+  var tempPointer = result;
+  var conn:tcpConn;
+  while tempPointer != nil {
+    var address = new ipAddr(tempPointer.addr);
+    try {
+      conn = connect(address, timeout);
+      break;
+    }
+    catch {
+      tempPointer = tempPointer.next;
+      continue;
+    }
+  }
+  sys_freeaddrinfo(result);
+  if tempPointer == nil {
+    throw new Error("Can't resolve address");
+  }
+  return conn;
+}
+
+/*
+  This overload of `connect` not only returns a :type:`tcpConn`
+  but also does DNS resolution for the provided `host`.
+  The `timeout` is tired for all resolved addresses and the first
+  successful one is returned back.
+
+  .. code-block:: Chapel
+
+    const timeout = new timeval(4,0);
+    const connectedClient = connect("google.com", 80, IPFamily.IPv4, timeout);
+
+  :arg host: address of host to connect or resolve if not in ip notation
+  :type host: `string`
+  :arg port: port to connect to on `host`
+  :type port: `uint(16)`
+  :arg family: type of socket family to connect to
+  :type family: :type:`IPFamily`
+  :arg timeout: time to wait for each possible connection.
+  :type timeval: :type:`~Sys.timeval`
+  :return: connected socket.
+  :rtype: `tcpConn`
+  :throws SystemError: Upon failure to resolve address or connect
+                    to any of the resolved address in given `timeout`.
+*/
+proc connect(in host: string, in port: uint(16), family: IPFamily = IPFamily.IPUnspec, in timeout = new timeval(-1,0)): tcpConn throws {
+  return connect(host, port:string, family, timeout);
+}
+
+/*
+  A record holding reference to a udp socket
+  bound to any available port.
+*/
+record udpSocket {
+  /*
+    File Descriptor Associated with instance
+  */
+  var socketFd: int(32);
+  /* Create a UDP socket of provided Family. */
+  proc init(family: IPFamily = IPFamily.IPv4) {
+    this.socketFd = -1;
+    try! this.socketFd = socket(family, SOCK_DGRAM);
+  }
```

Comment:
As with tcpListener, this record also needs a `deinit` which probably should close the socket.
It's important to still have a `close` method since `close` can fail (because there is no way of getting the returned error out of a `deinit`).